### ISSUE TITLE
esbuild should move to version 1.0.0

### ISSUE THREAD

**Issue by dev-ardi** [evanw/esbuild#2544]
Based on https://semver.org/#how-do-i-know-when-to-release-100

I feel like the API is *stable enough* across versions, and it's used in production in a vast number of projects, companies and packages.

As a user of esbuild API I feel like it's important for the version number to comunicate how compatible one version is from the next: Can I safely update my esbuild in my project without having to read all of the patch notes?

**Comment by evanw** [comment]
> Can I safely update my esbuild in my project without having to read all of the patch notes?

First, here's what esbuild currently does:

```
0.MINOR.PATCH
    │     │
    │     ╰─ Updating is probably safe (outside of some obscure edge cases)
    ╰─ Upgrading is probably not safe (a reasonable number of people may be impacted)
```

This issue is a request to change things to this instead (which is what semver is):

```
MAJOR.MINOR.PATCH
  │     │     │
  │     ╰─────┴─ Updating is totally safe
  ╰─ Upgrading could hypothetically not be safe
```

I've been thinking about this and I'm not sure how I feel about it. I agree that esbuild is used widely in production, and I do make an effort to keep esbuild stable unless a change is necessary. But esbuild is also very complex with a large API surface and many interconnected features, so many changes that I often make to esbuild would require a major version bump for safety under this definition.

Here are some recent examples which I have considered to be patch changes but which would be semver major changes under this definition:

* Even something seemingly benign like accepting a new type of value in an API can cause breakage for someone when upgrading: https://github.com/davipon/esbuild-plugin-pino/issues/65. So this is an incompatible API change, which requires a major version bump under semver.

* Fixing a minification bug can change behavior: https://github.com/evanw/esbuild/issues/2361. This is not backwards compatible because it could result in a behavior change in bundled code that has come to rely on the old, buggy behavior. So this also requires a major version bump for safety.

* When I released support for Yarn Plug'n'Play, I needed to follow that up with some bug fixes. One such fix changed how paths were interpreted, which could cause a different file to be imported when a file has the same name as the enclosing directory: https://github.com/evanw/esbuild/issues/2455. The feature had just been released and was still pretty broken, so realistically no one was using it enough to be broken by this obscure bug fix. However, this bug fix is not backwards compatible because it could hypothetically result in a behavior change in bundled code, so this would have required a major version bump for safety.

* I recently added `names` to the source maps that esbuild generates: https://github.com/evanw/esbuild/issues/1296. Doing this could hypothetically change the behavior of tools that process esbuild's source maps, so this would necessitate a major version bump under semver.

* There are many of these examples...

Following "can I safely update esbuild in my project" rigidly basically makes semver meaningless for regularly-updated packages like esbuild, as there would quickly be a major version bump and your question will almost always be irrelevant (you just wouldn't get any updates at all).

One way of looking at this is to say "semver is supposed to be about API changes, so bug fixes that don't change the API are fine even if they could hypothetically break someone." But that doesn't accomplish your goal of "can I safely update esbuild in my project." It has also been clarified that by "API" semver means all behavior, not just the interface: https://github.com/semver/semver/issues/874#issuecomment-1247614295.

Another way of looking at this is to say "if a bug fix realistically isn't going to break 99.9% of people, then it can be a patch." But then that's not semver, which is problematic because npm has set up their ecosystem such that semver is expected. It also relies on opinions and guesses, which goes against the goal of using a version number as a guarantee.

> As a user of esbuild API I feel like it's important for the version number to comunicate how compatible one version is from the next

Ironically I think this is what makes semver not a good fit for esbuild. I feel that the current versioning approach more effectively comunicates how compatible one version is from the next than semver. With the current approach the answers are "probably not" or "probably" but with semver the answers are "definitely not" or "perhaps" (with most updates falling under "perhaps"). Semver makes it hard to distinguish widely-compatible from likely-incompatible changes.

**TL;DR:** With the current versioning scheme, you should pin esbuild to a specific version and not update it unless you need to. If you want to upgrade esbuild to version "0.x.y" to fix a specific bug you're hitting or to get a new feature that you need, you can expect minimal breakage when upgrading if "x" stays the same but you should be aware that things might break when upgrading if "x" changes.

**Comment by shellscape** [comment]
I'm generally fine with the 0.x. I know exactly where I stand.

**Comment by manast** [comment]
There are a lot of valid points in this post, although I think it is actually possible to have and eat the cake here, because basically you can just have a versioning system that is close to SemVer in spirit but that fits your particular view of versioning. For instance, major versions increase when the risk of breaking something is pretty high, micro versions increase as today, only for bug fixes, minor for new features with a very high confidence that they would not break anything, otherwise increase major. That would keep most people happy I think, and also would remove that 0 for major which looks a bit useless right now and just seems to signal that this is not a production ready product.

**Comment by rtritto** [comment]
With major version 0, each library use a different `esbuild` version (no dedupe dependencies).

```console
➤ YN0000: · Yarn 4.4.1
➤ YN0000: ┌ Resolution step
➤ YN0000: └ Completed
➤ YN0000: ┌ Fetch step
➤ YN0000: └ Completed
➤ YN0000: ┌ Link step
➤ YN0007: │ esbuild@npm:0.23.1 must be built because it never has been before or the last one failed
➤ YN0007: │ esbuild@npm:0.21.5 must be built because it never has been before or the last one failed
➤ YN0007: │ esbuild@npm:0.19.12 must be built because it never has been before or the last one failed
➤ YN0000: └ Completed
➤ YN0000: ┌ Deduplication step
➤ YN0000: │ No packages can be deduped using the highest strategy
➤ YN0000: └ Completed
➤ YN0000: · Done
```

If `esbuild` is stable, it's better to upgrade to v1.

FYI @evanw

**Comment by jakebailey** [comment]
Bumping to v1 would not help that. v0.22 to v0.23 contained breaking changes, which is how it works for 0.x bumps. Moving to v1 would just mean that the same breaks would cause a bump to v2, causing the same duplication in `node_modules` (because they are not compatible).

**Comment by mark-wiemer** [comment]
@evanw it's been nearly two years and about 8 bumps (from 0.16.0 to 0.24.0), have you reconsidered this? For me it was a bit of a turnoff as I thought this project was still early in development or significantly unstable. It also feels strange not knowing whether a new release includes just bugfixes (v1.0.1) or new features (v1.1.0).

That said, I do think you should stick to semantic versioning regardless, and staying in v0 is OK by me. I've seen enough projects rely on this and looked at the changelog enough to know this is a stable project ready for production use :)

**Comment by casewalker** [comment]
In my opinion, staying on version zero forever and having breaking changes between patch versions is not the right way to run a longterm project.

Without getting to major version 1, there is no way to fix vulnerabilities in past versions (such as [GHSA-67mh-4wv8-2f99](https://github.com/advisories/GHSA-67mh-4wv8-2f99)) without breaking semver. Anyone who wants to avoid the vulnerability then needs figure out all of the breaking changes lying between their version and the stable version.

Thank you for this wonderful project, but I strongly urge you to change your versioning strategy.

**Comment by hyrious** [comment]
> Without getting to major version 1, there is no way to fix vulnerabilities in past versions (such as [GHSA-67mh-4wv8-2f99](https://github.com/advisories/GHSA-67mh-4wv8-2f99?rgh-link-date=2025-02-13T03%3A15%3A14.000Z)) without breaking semver.

It actually can. For example if you want to backport a fix in `0.14.54`, just publish a `0.14.55` which is updateable from the `^0.14.54` specifier.

**Comment by casewalker** [comment]
@hyrious No, that's not possible if you have already published a version `0.14.55`. Being in version zero (0.y.z) means [anything can change at any time](https://semver.org/#spec-item-4), so there is no guarantee that there weren't breaking changes between `0.14.54` and `0.14.55`, so even if a patch was published on the next available `0.14.z` version, it is not guaranteed that normal semver rules apply.

Also the `^0.14.54` specifier you mention would pick up all minor and patch changes, so if version `0.15.0` had been published and there were breaking changes, you would pick those up too.

**Comment by jakebailey** [comment]
@casewalker You're mistaken about how semver works with `0.X.Y` versions; `^0.14.0` will match `0.14.0`, `0.14.1`, etc, but will not match `0.15.0`. Having a zero as major version means the minor version is treated as the major version for purposes of `^` matching. Try searching for `esbuild@^0.14.0` on https://semver.npmjs.com.

**Comment by mark-wiemer** [comment]
@casewalker you're right that anything can change at any time, but `^0.14` will not pick up `0.15`, that's the beauty of major version zero. `^` looks at the first non-zero number and ensures it's fixed. Ref https://semver.npmjs.com/#syntax-examples

**Comment by casewalker** [comment]
Ah, @mark-wiemer @jakebailey my mistake, I'll remove that part.

**Comment by mark-wiemer** [comment]
Welcome to open-source!

![dependency_2x](https://github.com/user-attachments/assets/2c7b8b33-c9bf-46b3-a5de-16b629dbd03a)

https://xkcd.com/2347/

My advice: know your dependencies, know the severity of failures, remember that you can always pin versions to avoid regressions or malicious attacks, and cross your fingers!

With esbuild for my use-cases specifically, as long as my tests pass, I'm confident esbuild did its job :)

**Comment by casewalker** [comment]
@mark-wiemer That is a sadly true and favorite XKCD.

I want to point out that my argument hinges on the fact that you **cannot** pin versions to avoid malicious attacks. Generally speaking, any dependency can be found to have a vulnerability at any time, and if you stay pinned to a vulnerable version, you are almost asking to be attacked. From that, I am arguing that longterm, stable projects staying in v0 forever removes potential paths to remediating that vulnerability (unless you are willing to break SemVer).

**Comment by manast** [comment]
@casewalker you are rising a super important point here regarding security. I think it is not wise to just rebel from the best practices and standards that have been proven to work reasonably well during more than a decade now, just because you think your project is different, or that it has some particularities that makes it not need to follow the recommended guidelines... it's frustrating because, in the end, this kind of approach can lead to real harm.

**Comment by GabenGar** [comment]
That's a cool comic but doesn't answer the question on how to resolve the dependency issues caused by CVE mentioned above.
Currently `npm audit` tells to downgrade [`@vitejs/plugin-react`](https://github.com/vitejs/vite-plugin-react) to [`1.3.2`](https://github.com/vitejs/vite-plugin-react/commit/7049973f22b7b165fd69183051fc67453f101e3b), which is 3 major versions behind and 3 years ago (that's basically prehistoric age in react land). I am pretty sure it will just introduce more vulnerabilities.

**Comment by terrablue** [comment]
My two cents here: don't ever release major 1. At least, don't do it for a *technical reason*.

esbuild, as far as I'm concerned, is rock-solid. I've been using it for a long time, in production. Moving to major 1 would be just an unnecessary declaration of confidence that should have -- and has already -- been established by the project's track record. It's the classic actions speak louder than words.

I can only see downsides to moving to major 1. Suddenly, you need to think whether this change is a minor or patch. This isn't relevant to most users, and it's certainly irrelevant to npm on `^`, which is what almost everyone uses.

`0.y.z` keeps it simple. A y-upgrade is potential breakage, and a z-upgrade is safe. It's reducing semver to its core, breakage vs. non-breakage, and that is all that matters.

**Comment by manast** [comment]
@terrablue your same argument could be applied to basically any npm package, what makes esbuild special so that it should not adhere to the best practices and standards in the NodeJS ecosystem?

**Comment by terrablue** [comment]
@manast I was not making an esbuild-specific argument. IMHO, major 1 is mostly public relations. Bun, for example, is on major 1 whilst relying on Zig, which is on 0 and will remain so for the foreseeable future as far as I can tell. Also, it still has its fair share of bugs and unfinished features, which is to be expected of a relatively new project. All in all, it's quite arguably *less* stable than esbuild, and yet it's on major 1. It's *not* a technical decision.

Also, major 1 is *meaningless* if the project has no good track record. IOW, anyone could publish a major 1, and that is no guarantee for stability.

I was arguing that esbuild should continue do what it's doing -- essentially work with a minimal subset of semver that is both realistic and simple. It has served the project well and there is no reason to change it.

**Comment by manast** [comment]
> Yeah, Bun, Zig, esbuild or whatever everyone should use 0.x forever. Semver is a mistake, it doesn't work. There's no reason to use major 1. It's pointless as you said. 🗯️

How to say I never worked in production code without saying I never worked in production code 😂

**Comment by GabenGar** [comment]
@terrablue 
> Bun, for example, is on major 1 whilst relying on Zig, which is on 0 and will remain so for the foreseeable future as far as I can tell. Also, it still has its fair share of bugs and unfinished features, which is to be expected of a relatively new project. All in all, it's quite arguably less stable than esbuild, and yet it's on major 1. It's not a technical decision.

Cool story, but also completely missing a point that 1.0.0 isn't some stability or feature complete indicator.

**Comment by shellscape** [comment]
@evanw time to convert this to a discussion. I have no idea why anyone is still bikeshedding this.

**Comment by lazarljubenovic** [comment]
Don't _ever_ close this issue; it's an endless source of entertainment.

**Comment by PeterJSung** [comment]
Releasing semversion 1 isn't an unnecessary display of confidence. Typically, when users install a package and see a major version number of 0, they assume it's a beta version, a development version that hasn't been officially released yet.

It's true that there's been a lot of discussion about updating from version 0 to version 1.
https://stackoverflow.com/questions/44797849/when-using-semver-when-to-upgrade-bump-to-1-0-0-stable

According to semver org, version 0.x.y is the initial release, and this is something that many developers are aware of.
https://semver.org/#spec-item-4

At least the build related features are complete and I think it should be upgraded to version 1 unless there are some major modifications.

**Comment by casewalker** [comment]
> `0.y.z` keeps it simple. A y-upgrade is potential breakage, and a z-upgrade is safe. It's reducing semver to its core, breakage vs. non-breakage, and that is all that matters.

@terrablue No, that's false, that's not how [SemVer is defined](https://semver.org/):
> 4 . Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.
> ...
> 6 . Patch version Z (x.y.Z | x > 0) MUST be incremented if only backward compatible bug fixes are introduced. A bug fix is defined as an internal change that fixes incorrect behavior.

Patch versions are defined specifically to only have meaning for code which has released major version >0. Anything is allowed to break at any time while still in major version zero.

**Comment by terrablue** [comment]
@casewalker I think you may have missed my point. For 0.y.z packages, given you use ^, which is what most people do, npm will install any patch updates, but not any minor updates. That's my "reduce semver to breaking vs non-breaking".
